debian/cloud-init.postinst

```diff
@@ -276,6 +276,51 @@ fix_azure_upgrade_1611074() {
     rm "$tmpf" || :
 }
 
+fix_lp1889555() {
+    local oldver="$1" last_bad_ver="20.2-94-g3d06abc2-0ubuntu1"
+    dpkg --compare-versions "$oldver" le-nl "$last_bad_ver" || return 0
+
+    # if cloud-init's grub module did not run, then it did not break anything.
+    [ -f /var/lib/cloud/instance/sem/config_grub_dpkg ] || return 0
+
+    # Don't do anything unless we have grub
+    [ -x /usr/sbin/grub-install ] || return 0
+
+    # Make sure that we are not chrooted.
+    [ "$(stat -c %d:%i /)" != "$(stat -c %d:%i /proc/1/root/.)" ] && return 0
+
+    # Check if we are in a container, i.e. LXC
+    if systemd-detect-virt --quiet --container || lxc-is-container 2>/dev/null; then
+        return 0
+    fi
+
+    # This bug only applies to NVMe devices
+    [ -e /dev/nvme0 ] || return 0
+
+    db_get grub-pc/install_devices && grub_cfg_dev=${RET} || return 0
+
+    # If the current setting is not the (incorrect) default we expect, this
+    # implies user intervention so leave things alone
+    # XXX: Do we thinkg this is the correct behaviour?
+    [ "$grub_cfg_dev" = "/dev/sda" ] || return 0
+
+    correct_idev="$(python3 -c "import logging; from cloudinit.config.cc_grub_dpkg import fetch_idevs; print(fetch_idevs(logging.getLogger()))")" || return 0
+
+    # If the correct_idev is already configured, do nothing
+    [ "$grub_cfg_dev" = "$correct_idev" ] && return 0
+
+    echo "Reconfiguring grub install device due to mismatch (LP: #1889555)"
+    echo "   grub should use $correct_idev but is configured for $grub_cfg_dev"
+    db_set grub-pc/install_devices "$correct_idev"
+
+    if grub-install "$correct_idev"; then
```

Comment:
@vorlonofportland @xnox I feel like we never fully bottomed out on this discussion: should the cloud-init postinst perform a `grub-install` once it has set `grub-pc/install_devices` to be correct, to test that the new value can actually be installed to? Or should it just perform the configuration and leave it at that?

(This pattern is used in another postinst snippet which also handles grub misconfiguration: https://github.com/canonical/cloud-init/blob/ubuntu/devel/debian/cloud-init.postinst#L193-L199, which is why I followed it here.)

Reply:
i think it should

Reply:
grub-install is a risky operation, and we are specifically not calling it for updating the BIOS bootloader on ANY systems within the SRU stream, only on release-upgrades.  So cloud-init should also not call grub-install on upgrade.

Reply:
Right, I too change my mind. cloud-init must not call grub-install. Instead, I wish this cloud-init sru to land, such that grub2 sru can declare breaks on the broken versions of cloud-init and therefore ensure that upon distr-upgrade cloud-init is configured first (and fixes up the install-devices), such that when grub2-pc is configured, it can safely call grub-install with the correct failure condition handling and recovery.